Use a pool of workers which cache connections

ash-owl opened the pull request:
At the moment, the LDAP backend opens a connection to the LDAP server for each query.

<del>This adds a pool of workers which can each hold a bit of state; queries are done on the workers and the state is used to cache the LDAP connections.</del>


<del>The supervision tree ends up looking like this:
![screenshot - 060315 - 17 48 20](https://cloud.githubusercontent.com/assets/6576021/6530465/090f5640-c429-11e4-85cf-d71e3779d4f1.png)</del>

<del>`rabbit_auth_backend_ldap_pool_coord` is a `gen_server` which gets called by both:</del>

- <del>clients (processes that want work to be done)</del>
- <del>workers (processes that do the work)</del>

<del>...and matches them up.</del>


<del>Doing it this way (having the workers call the coordinator rather than the coordinator call the workers) makes the workers and supervisor pretty simple. The coordinator is somewhat less simple, but still fairly simple.</del>


EDIT: removed the description of the now-removed worker pool.

ash-owl committed:
use a pool of workers which cache connections

ash-owl commented on src/rabbit_auth_backend_ldap_app.erl:
```diff
@@ -50,4 +50,18 @@ configured(M,  [_    |T]) -> configured(M, T).
+               worker,
+               [rabbit_auth_backend_ldap_pool_coord]},
+  WorkerSupSpec ={rabbit_auth_backend_ldap_pool_worker_sup,
+                  {rabbit_auth_backend_ldap_pool_worker_sup, start_link, [10, {3, 10}]},
```
should make the number of workers configurable

ash-owl commented on src/rabbit_auth_backend_ldap_pool_coord.erl:
```diff
@@ -0,0 +1,188 @@
+      try
+        gen_server:reply(Worker, Work)
+      catch
+        E -> io:format("exception: ~p~n", [E])
```
need to deal with that properly

ash-owl commented on src/rabbit_auth_backend_ldap_pool_coord.erl:
```diff
@@ -0,0 +1,188 @@
+  check_invariants(State1),
+  State1.
+
+check_invariants(State=#state{idle_workers=IdleWorkers,
```
should probably be a compile-time option

ash-owl commented on src/rabbit_auth_backend_ldap_pool_coord.erl:
```diff
@@ -0,0 +1,188 @@
+      try
+        R = gen_server:reply(Client, {error, worker_failed})
+      catch
+        E -> io:format(user,"exception: ~p~n", [E])
```
erk

ash-owl commented on test/src/rabbit_auth_backend_ldap_pool_coord_test.erl:
```diff
@@ -0,0 +1,88 @@
+
+busy_worker_fails() ->
+  Worker = spawn(fun () ->
+                     ?debugHere,
```
noise

michaelklishin commented:
Thank you, connection pooling is definitely a good idea!

simonmacmullen commented:
OK, first of all: thanks for this!

So I'd like to talk a bit about the design. Firstly, let's talk about how the existing LDAP plugin performs. It always establishes a new LDAP connection, then binds, then performs any necessary queries. Notably it does this for `vhost_access_query` and `resource_access_query`, even if the "query" is `{constant, whatever}`: it still establishes a connection and binds, then does nothing. (Assuming you are not using `{other_bind, anon}` in which case it skips the bind). I added some trivial benchmarking to test a simple `vhost_access_query` over a connection with a latency of 10ms. Unsurprisingly the time taken was a little over 60ms - that's three round trips for the connection establishment, binding, and query.

(See https://gist.github.com/simonmacmullen/0e4c88cc3ffed2f74f3a for a bit more on performance testing.)

Now for your branch - it reduces this to a little over 40ms, since it skips the TCP connection establishment, but still binds. Obviously with `{other_bind, anon}` it does better. For a `{constant, whatever}` query it goes from 40ms to 20ms. However, if we could skip having to repeatedly rebind, we could get that down to ~20ms and ~0ms respectively.

So: could the worker processes keep knowledge of what creds they last bound as, and skip the bind if they have not changed?

Of course, now you have a cache invalidation problem: if the workers keep hold of creds, what happens when an admin locks someone out via LDAP? We need to not hold onto the creds forever. So the workers could send themselves a message when they bind, saying "expire this binding after (say) 5 minutes".

Of course, if you went down this path, it would be nice if you could avoid churning the LDAP connections' bindings when multiple users are connecting to the server. One way to do this would be to have one worker process per user.

In fact, do you really need the connection pool part of this at all? Why not just have one worker process per user, owning one LDAP connection, have that expire completely after five minutes (in case the password changes, and to stop leaks when there's a big turnover of users). Have some ETS table to look them up in. Done. Yes, you could have lots of LDAP connections, but only if you have a vast number of users and all those users connect simultaneously (in which case you have a vast number of AMQP connections too and in general need to tune for lots of connections).

However, you might not want to rework it all. I would suggest that what I've just outlined doesn't need to be any more work than what you've already produced, and I would also suggest that at least having some way to detect and skip rebinding is likely to be very useful, the performance gains should be quite substantial. But I'll move onto what else we could accept in another comment.

simonmacmullen commented:
OK, let's assume you don't want to rewrite everything as the previous comment suggested. This branch still gives at least some improvement in performance, so what's the minimum to do to get it merged?

There are a couple of issues. One is that the branch invents a new worker pool mechanism. The RabbitMQ server already has a battle-tested and debugged worker pool - see `worker_pool.erl`. I would be much happier seeing this branch use that code than bring in another worker pool implementation.

Now the existing `worker_pool` would need some modification, but it could be quite minor. At the moment it establishes a single pool with a hard-coded name; really you want an LDAP-specific pool. So `worker_pool` would need to allow you to create and use named pools. That's only a handful of lines of code to change though. You would also need to add some way for the workers to maintain state. Or you could just stick the state in the process dictionary :-)

A second issue that I found is that this branch can deadlock under load. To see this happen the easiest thing to do is hard code the pool size to 1, and then run the tests. You'll see some of the tests time out because they deadlocked. (`login_test_()` with `other_bind_broken_env()` timed out for me).

I'm pretty sure this is happening because here: https://github.com/ash-lshift/rabbitmq-auth-backend-ldap/blob/bdf08ef8823d06c265e7b2a05d6525d81aad9a6d/src/rabbit_auth_backend_ldap.erl#L371 we call into `with_ldap/2` when we are already within `with_ldap/2`. So either we need to stop doing that, or `with_ldap/2` needs to become re-entrant somehow. I would guess that the former is probably going to be easier.

ash-owl commented:
Actually, the motivation for this change was a system with many, many connections and many, many users :) So my goals here are
- as good or better performance in all cases
- significantly better performance in the many connections, many users case

Hopefully this is a defensible position.

Detecting and skipping rebinds is certainly a tempting low-hanging fruit. I didn't bother with it here because just pooling the connections turned out to be enough of an improvement for the system in question.

Well spotted that `with_ldap` isn't re-entrant. I'll sort that out.

I completely agree that re-using the existing worker pool would be much less nerve-wracking, and I had considered doing so. I can't remember why I decided against it now, so let me look into it again.

simonmacmullen commented:
I think it's certainly defensible. The more you can do to move in the direction of skipping rebinds and having some worker-user affinity the happier I'll be, but ultimately this would still be mergeable with neither of those.

ash-owl committed:
use worker_pool from rabbitmq-server

...instead of the NIH worker pool implementation. This assumes an as-yet
hypothetical modification to rabbitmq-server.

ash-owl commented:
Good call on reusing the worker pool. It turned out to be straightforward to make the necessary changes : see [here](https://github.com/rabbitmq/rabbitmq-server/compare/master...ash-lshift:reusable-worker-pool?expand=1) for a diff. I haven't had a chance to work up a proper PR for that yet but thought I'd show you all what it looks like here before I take it any further.

IIRC `worker_pool` already takes care of non-reentrant work.

By the way, can anyone give me a hint about how things will progress if you decide that you want to accept the change? Is there a staging branch for 3.6.0 or will I have to track the changes in 3.5.x for example?

michaelklishin commented:
@ash-lshift master will become `3.6.0`.

Commit:
Make LDAP worker pool size configurable.

alexethomas commented:
PR rabbitmq/rabbitmq-server#83 created for the worker pool change so I think this is complete. 
Well, complete except another stanza to the LDAP doc page is needed for the new <code>pool_size</code> parameter:

pool_size

```
LDAP worker pool size. Default: 10.

This pool manages the connections made to the LDAP server. Each worker maintains two connections, one anonymous and one for binding.
```

ash-owl commented:
(For the avoidance of doubt: @alexethomas has kindly adopted this PR from me since I'm tied up with other things ATM.)

Commit:
Merge branch 'master' into pull-request-83-fixups

Commit:
Minor cleanups
Factor out the rebinding bit to stop the function getting out of hand,
reinstate a TODO since while better it is still not perfect, clean up
a few indent issues to conform to our poorly documented code standards.

simonmacmullen commented:
I've created a branch `pull-request-4-fixups` in our repo to clean up the code a little. Hopefully it will not be too onerous for you to merge this in to your repo and we can go back and forth that way.

So this still isn't there I'm afraid. It's much closer, but there's a nasty issue: it both allows LDAP connections to be long lived and does not deal with failed ones.

To test this: set your `pool_size` to 1, log in, then restart your LDAP server. Logins will now break until you restart RabbitMQ.

There's also a small race condition at startup: the `auth_backend` depends on the worker pool. But we might start using the `auth_backend` as soon as RabbitMQ (or rather, the `rabbit` app) accepts TCP connections, but the worker pool starts later than that, with the `rabbitmq_auth_backend_ldap` app.

So the worker pool should be started in a `boot_step`, not by the app. Federation does this for its link process supervision hierarchy, so https://github.com/rabbitmq/rabbitmq-federation/blob/master/src/rabbit_federation_sup.erl should be inspirational.

Commit:
Recover from closed connection to server.

Commit:
Move LDAP pool creation from app init to a boot step.

Commit:
Move connection error detection up a level to catch anon access error.

alexethomas commented:
Hi Simon, thanks very much for the guidance. We've followed all the above so hopefully getting near the mark now.

michaelklishin commented:
LGTM, tests pass, manual testing reveals no breakage and double digit % throughput gains on basic connection stress tests.

We'll be doing several preview releases of 3.6.0 to let more people test this extensively. It would be great if we could cut down on extra bind requests before 3.6.0 but this is already an improvement.

michaelklishin commented:
@ash-lshift @alexethomas thank you for working on this!

ash-owl commented:
Thanks @michaelklishin and @simonmacmullen !
